File: AlgorithmsProject/src/in/designpatterns/observer/WeatherData.scala
package in.designpatterns.observer

class WeatherData extends Observable {
  var temperature: Float = _
  var humidity: Float = _
  var pressure: Float = _

  def setMeasurements(temp: Float, hum: Float, press: Float) = {
    this.temperature = temp
    this.humidity = hum
    this.pressure = press
    measurementsChanged
  }

  def measurementsChanged = {
    setChanged
    notifyObservers
  }
}

File: AlgorithmsProject/src/in/designpatterns/scala/observer/WeatherData.java
package in.designpatterns.scala.observer;

import java.util.ArrayList;
import java.util.List;

public class WeatherData implements Observable {
    private final List<Observer> observers = new ArrayList<>();
    private float temperature;
    private float humidity;
    private float pressure;

    @Override
    public void registerObserver(Observer o) {
        observers.add(o);
    }

    @Override
    public void removeObserver(Observer o) {
        observers.remove(o);
    }

    @Override
    public void notifyObservers() {
        for (Observer o : observers) {
            o.update(temperature, humidity, pressure);
        }
    }

    public void setMeasurements(float temp, float hum, float press) {
        this.temperature = temp;
        this.humidity = hum;
        this.pressure = press;
        notifyObservers();
    }
}
